src/utils.js

```diff
@@ -66,9 +76,342 @@ export function bytesAreEqual(v1, v2) {
+// of the buffer, as it indicates a malformed TLS message of some sort.
+
+class BufferWithPointer {
+  constructor(buf) {
+    assertIsBytes(buf);
+    this._buffer = buf;
+    this._dataview = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
+    this._pos = 0;
+  }
+
+  resize(size) {
+    assert(size > this.length(), 'cant resize BufferWithPointer to be smaller');
+    const newbuf = new Uint8Array(size);
+    newbuf.set(this._buffer.slice(0, this.tell()), 0);
+    this._buffer = newbuf;
+    this._dataview = new DataView(newbuf.buffer, newbuf.byteOffset, newbuf.byteLength);
+  }
+
+  length() {
+    return this._buffer.byteLength;
+  }
+
+  tell() {
+    return this._pos;
+  }
+
+  seek(pos) {
+    this._pos = pos;
+    assert(this._pos <= this.length(), 'do not seek past end of buffer');
+  }
+
+  incr(offset) {
+    this.seek(this._pos + offset);
+  }
+
+  slice(start = 0, end = this.tell()) {
+    return this._buffer.slice(start, end);
+  }
+}
+
+
+export class BufferReader extends BufferWithPointer {
+
+  hasMoreBytes() {
+    return this.tell() < this.length();
+  }
+
+  readBytes(length) {
+    const slice = this.slice(this.tell(), this.tell() + length);
+    this.incr(length);
+    return slice;
+  }
+
+  readRemainingBytes() {
+    return this.readBytes(this.length() - this.tell());
+  }
+
+  readUint8() {
+    const n = this._dataview.getUint8(this._pos);
+    this.incr(1);
+    return n;
+  }
+
+  readUint16() {
+    const n = this._dataview.getUint16(this._pos);
+    this.incr(2);
+    return n;
+  }
+
+  readUint24() {
+    let n = this._dataview.getUint16(this._pos);
+    n = (n << 8) | this._dataview.getUint8(this._pos + 2);
+    this.incr(3);
+    return n;
+  }
+
+  readUint32() {
+    const n = this._dataview.getUint32(this._pos);
+    this.incr(4);
+    return n;
+  }
+
+  //
+  // These are helpers for reading the variable-length vector structure
+  // defined in https://tools.ietf.org/html/rfc8446#section-3.4.
+  //
+  // Such vectors are represented as a length followed by the concatenated
+  // bytes of each item, and the size of the length field is determined by
+  // the maximum allowed size of the vector.  The size of the length field
+  // is determined by the maximum number of bytes in the vecctor.  For example
+  // to read a vector that may contain up to 65535 bytes, use `readVector16`.
+  //
+  // To read a variable-length vector of between 1 and 100 uint16 values,
+  // defined in the RFC like this:
+  //
+  //    uint16 items<2..200>;
+  //
+  // You would do something like this:
+  //
+  //    const items = []
+  //    buf.readVector8(buf => {
+  //      items.push(buf.readUint16())
+  //    })
+  //
+
+  _readVector(length, cb) {
+    const limit = this.tell() + length;
+    // Keep calling the callback until we've consumed the expected number of bytes.
+    // It can return `true` to indicate we should stop iterating and skip the remainder.
```

Comment:
I would disable this feature.  The callback should be run once for each item in the list until it is done.

One thing I would do with the callback is pass it a different instance of this object that offers a view onto the same buffer, but with a the shorter length.  Then they can't read past "length" by design and you don't have to worry about inner reads going outside of the space that is available.